File: courier/src/test/format.test.ts
import { describe, expect, it } from "vitest";
import {
  money,
  formatDateTime,
  formatDay,
  statusLabel,
  statusPill,
  paymentLabel,
  mapsUrl,
  qtyUnit,
  distanceLabel,
  etaLabel,
} from "../lib/format";

/** toLocaleString ru-RU non-breaking space (\u00a0) ni oddiy space ga almashtiradi. */
const NBSP = "\u00a0";

describe("money", () => {
  it("formats number with spaces", () => {
    expect(money(1000)).toBe(`1${NBSP}000`);
    expect(money(1234567)).toBe(`1${NBSP}234${NBSP}567`);
    expect(money(0)).toBe("0");
  });
});

describe("formatDateTime", () => {
  it("formats ISO string", () => {
    const r = formatDateTime("2026-06-30T12:00:00Z");
    expect(r).toContain("30.06");
  });
});

describe("formatDay", () => {
  it("formats as day + short month", () => {
    const r = formatDay("2026-06-30");
    expect(r).toContain("30");
  });
});

describe("statusLabel", () => {
  it("returns label for known status", () => {
    expect(statusLabel("pending")).toBe("Yangi");
    expect(statusLabel("accepted")).toBe("Qabul qilindi");
    expect(statusLabel("delivering")).toBe("Yetkazilmoqda");
    expect(statusLabel("delivered")).toBe("Yetkazildi");
    expect(statusLabel("cancelled")).toBe("Bekor qilindi");
  });

  it("falls back to the status value for unknown", () => {
    expect(statusLabel("unknown" as any)).toBe("unknown");
  });
});

describe("statusPill", () => {
  it("returns CSS class for known status", () => {
    expect(statusPill("delivered")).toContain("emerald");
    expect(statusPill("cancelled")).toContain("rose");
    expect(statusPill("accepted")).toContain("cyan");
  });

  it("falls back to default for unknown", () => {
    expect(statusPill("unknown" as any)).toContain("slate");
  });
});

describe("paymentLabel", () => {
  it("returns Uzbek label", () => {
    expect(paymentLabel("cash")).toBe("Naqd");
    expect(paymentLabel("payme")).toBe("Payme");
  });

  it("returns dash for undefined", () => {
    expect(paymentLabel(undefined)).toBe("—");
  });
});

describe("mapsUrl", () => {
  it("generates Yandex Maps URL", () => {
    const url = mapsUrl(41.3, 69.2);
    expect(url).toContain("yandex.com/maps");
    expect(url).toContain("rtext=~41.3,69.2");
  });

  it("returns null if lat/lng missing", () => {
    expect(mapsUrl(null, null)).toBeNull();
    expect(mapsUrl(undefined, undefined)).toBeNull();
  });
});

describe("qtyUnit", () => {
  it("uses provided unit", () => {
    expect(qtyUnit(2, "kg")).toBe("2 kg");
    expect(qtyUnit(1, "dona")).toBe("1 dona");
  });

  it("defaults to 'dona' when unit missing", () => {
    expect(qtyUnit(5, null)).toBe("5 dona");
    expect(qtyUnit(3, undefined)).toBe("3 dona");
  });
});

describe("distanceLabel", () => {
  it("shows km for >= 1 km", () => {
    expect(distanceLabel(4.2)).toBe("4.2 km");
    expect(distanceLabel(1.0)).toBe("1.0 km");
  });

  it("shows meters for < 1 km", () => {
    expect(distanceLabel(0.85)).toBe("850 m");
    expect(distanceLabel(0.05)).toBe("50 m");
  });

  it("returns null for null/undefined", () => {
    expect(distanceLabel(null)).toBeNull();
    expect(distanceLabel(undefined)).toBeNull();
  });
});

describe("etaLabel", () => {
  it("formats with ~ prefix", () => {
    expect(etaLabel(25)).toBe("~25 daqiqa");
    expect(etaLabel(10)).toBe("~10 daqiqa");
  });

  it("returns null for null/undefined/0", () => {
    expect(etaLabel(null)).toBeNull();
    expect(etaLabel(undefined)).toBeNull();
  });
});

File: kuryer/test/format_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:kuryer/core/format.dart';

/// Dart port of the courier web `src/test/format.test.ts` +
/// `orderActions.test.ts`, so the shared formatting/logic stays in lock-step.

const _acceptable = {'pending', 'confirmed', 'preparing', 'ready'};
bool isAcceptableOrderStatus(String s) => _acceptable.contains(s);

void main() {
  group('money', () {
    test('formats number with spaces', () {
      expect(money(1000), '1 000');
      expect(money(1234567), '1 234 567');
      expect(money(0), '0');
    });
  });

  group('formatDateTime', () {
    test('formats ISO string', () {
      expect(formatDateTime('2026-06-30T12:00:00Z'), contains('.06'));
    });
  });

  group('formatDay', () {
    test('formats as day + short month', () {
      expect(formatDay('2026-06-30'), contains('30'));
    });
  });

  group('statusLabel', () {
    test('returns label for known status', () {
      expect(statusLabel('pending'), 'Yangi');
      expect(statusLabel('accepted'), 'Qabul qilindi');
      expect(statusLabel('delivering'), 'Yetkazilmoqda');
      expect(statusLabel('delivered'), 'Yetkazildi');
      expect(statusLabel('cancelled'), 'Bekor qilindi');
    });

    test('falls back to the status value for unknown', () {
      expect(statusLabel('unknown'), 'unknown');
    });
  });

  group('paymentLabel', () {
    test('returns Uzbek label', () {
      expect(paymentLabel('cash'), 'Naqd');
      expect(paymentLabel('payme'), 'Payme');
    });

    test('returns dash for null', () {
      expect(paymentLabel(null), '—');
    });
  });

  group('mapsUrl', () {
    test('generates Yandex Maps URL', () {
      final url = mapsUrl(41.3, 69.2);
      expect(url, contains('yandex.com/maps'));
      expect(url, contains('rtext=~41.3,69.2'));
    });

    test('returns null if lat/lng missing', () {
      expect(mapsUrl(null, null), isNull);
    });
  });

  group('qtyUnit', () {
    test('uses provided unit', () {
      expect(qtyUnit(2, 'kg'), '2 kg');
      expect(qtyUnit(1, 'dona'), '1 dona');
    });

    test("defaults to 'dona' when unit missing", () {
      expect(qtyUnit(5, null), '5 dona');
      expect(qtyUnit(3, ''), '3 dona');
    });
  });

  group('distanceLabel', () {
    test('shows km for >= 1 km', () {
      expect(distanceLabel(4.2), '4.2 km');
      expect(distanceLabel(1.0), '1.0 km');
    });

    test('shows meters for < 1 km', () {
      expect(distanceLabel(0.85), '850 m');
      expect(distanceLabel(0.05), '50 m');
    });

    test('returns null for null', () {
      expect(distanceLabel(null), isNull);
    });
  });

  group('etaLabel', () {
    test('formats with ~ prefix', () {
      expect(etaLabel(25), '~25 daqiqa');
      expect(etaLabel(10), '~10 daqiqa');
    });

    test('returns null for null/0', () {
      expect(etaLabel(null), isNull);
      expect(etaLabel(0), isNull);
    });
  });

  group('isAcceptableOrderStatus', () {
    test('allows fresh and preparation-stage orders', () {
      expect(isAcceptableOrderStatus('pending'), true);
      expect(isAcceptableOrderStatus('confirmed'), true);
      expect(isAcceptableOrderStatus('preparing'), true);
      expect(isAcceptableOrderStatus('ready'), true);
    });

    test('hides accept action once claimed or finished', () {
      expect(isAcceptableOrderStatus('accepted'), false);
      expect(isAcceptableOrderStatus('delivering'), false);
      expect(isAcceptableOrderStatus('delivered'), false);
      expect(isAcceptableOrderStatus('cancelled'), false);
    });
  });
}
